### ISSUE TITLE
[Performance]: Custom fused kernel tracking

### ISSUE THREAD

**Issue by ProExpertProg** [vllm-project/vllm#25179]
### Proposal to improve performance

We need various additional fused kernels so we can allow fusion to happen orthogonally of other features. E.g. we want to support more kinds of quantization with existing passes, as well as allow for quantized kv cache to be orthogonal to fusion. Some of these kernels will require new passes, and most will require new patterns in existing passes.

### Flashinfer (B200)

Currently available (and integrated):
- silu_mul + nvfp4
- fp8 attention (decode) + static fp8 quant
- fp8 attention (prefill) + static fp8 quant
- fp8 attention (decode) + nvfp4 quant
- fp8 attention (prefill) + nvfp4 quant
- all_reduce + rms_norm
- all_reduce + rms_norm + static fp8 quant
- all_reduce + rms_norm + nvfp4 quant

Currently available (and not integrated):
- ? bf16 attention (decode) + static fp8 quant
- ? bf16 attention (decode) + nvfp4 quant
- rope + static fp8 quant 
- ? rope + kvcache 
- rope + static fp8 quant + kvcache
- ? mla attention (prefill) + static fp8 quant
- ? mla attention (prefill) + nvfp4 quant

I don't remember if bf16-attn with fused output quant was supported in prefill or decode but I think it was just one of the two. @pavanimajety is already working on integrating RoPE fusion. I also recall that mla prefill uses the mha kernel so this should be supported already but somebody also mentioned some cubins might need to be regenerated?

Needed (easy integration):
- rms_norm + nvfp4
- ? bf16 attention (prefill) + static fp8 quant
- ? bf16 attention (prefill) + nvfp4 quant
- all_reduce + rms_norm + dynamic per-token fp8 quant
- all_reduce + rms_norm + dynamic group fp8 quant

Needed (lower urgency):
- mla attention (decode) + static fp8 quant & nvfp4 quant
  - integration requires splitting decode/prefill out of MLA custom op and quantizing BMMs
- mla attention (prefill) + dynamic group fp8 quant
  - This is likely harder to fuse into attention but should be possible if attn output has `group_size` elements available at a time (up to 128)
- mla attention (prefill) + dynamic group fp8 quant
- all_reduce + rms_norm + dynamic per-tensor fp8 quant (low priority)

## Other hardware (Hopper, AMD, etc.)
I think the goal for expanded support should be custom fused Triton (or even Helion) kernels for better portability and to avoid combinatorial explosion, although if there are platform-specific kernels to be contributed they are certainly extremely welcome. Platform-specific might be required for collectives (or can we leverage Triton-Distributed?)

Current:
- attention + static fp8 quant fusion (Triton/ROCM backend)
  - we should make sure we maintain support for these in both the Triton and ROCm kernels (cc @tdoublep @bringlein)

Wanted:
- all_reduce + rms_norm (+ quant: all flavors)
- GEMM + static fp8 quant
  - Dynamic quant might be harder, again could be easier with group over per-token
  - Can we do this with our existing cutlass kernel epilogues?
- mla attention + quant
- aiter attention + quant

Wanted (only if faster than torch Inductor):
- rms_norm + quant
- silu_mul + quant

cc @pavanimajety @gshtras @kushanam @nvpohanh @mgoin @zou3519 @BoyuanFeng @tlrmchlsmth @LucasWilkinson 

### Your current environment (if you think it is necessary)

```text
The output of `python collect_env.py`
```

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by nvpohanh** [comment]
> 1. rope + static fp8 quant
> 2. ? rope + kvcache
> 3. rope + static fp8 quant + kvcache

These are in progress by @pavanimajety . (2) is just a special case of (3) and we should support that.

> 1. ? bf16 attention (decode) + static fp8 quant
> 2. ? bf16 attention (decode) + nvfp4 quant

I have been hesitant about this because supporting these means that we need to add a lot of new attn kernels into FlashInfer cubin, bloating the cubin sizes... @ProExpertProg Do we see use cases where users load FP8/FP4 models (which in some sense, they are okay with the accuracy drop) but still want to use FP16 kv-cache? If users care about performance, they should have used FP8 kv-cache, right?

Regarding mla, I will need @pavanimajety 's help since I am not that familiar with it.

**Comment by nvpohanh** [comment]
Regarding Hopper:

> all_reduce + rms_norm (+ quant: all flavors)

I think these FlashInfer kernels should be able to support SM90. We just need to modify that in FlashInfer's JIT flags.

**Comment by ProExpertProg** [comment]
> Do we see use cases where users load FP8/FP4 models (which in some sense, they are okay with the accuracy drop) but still want to use FP16 kv-cache? If users care about performance, they should have used FP8 kv-cache, right?

Currently 8-bit kvcache is opt-in as I assume it affects accuracy even though the rest of the model is quantized. But maybe it's not as common of a path to support at speed. cc @mgoin for decision here.

**Comment by mgoin** [comment]
I'm okay with preferring performance on Blackwell, so defaulting to fp8 kv cache to match fp8/fp4 activations. I think it especially makes sense when the checkpoint already has calibrated scales

Even with the fp8 attention in fa3 before, we weren't confident about the performance so it wasn't a clear win to choose

**Comment by ProExpertProg** [comment]
@nvpohanh what about these?
> - all_reduce + rms_norm + dynamic per-token fp8 quant
> - all_reduce + rms_norm + dynamic group fp8 quant

**Comment by nvpohanh** [comment]
> all_reduce + rms_norm + dynamic per-token fp8 quant
> all_reduce + rms_norm + dynamic group fp8 quant

I think all of these are doable, but it would be great if you can provide a list of models/configs that use this to justify the effort. i.e. let me know which models require it so that we can work on it and measure the performance gain from this.

**Comment by ProExpertProg** [comment]
> all_reduce + rms_norm + dynamic per-token fp8 quant

This is `llama-3.1-70B-FP8-dynamic` or `Llama-4-Scout-17B-16E-Instruct-FP8-dynamic`

> all_reduce + rms_norm + dynamic group fp8 quant

This is for Deepseek V3 / R1

**Comment by ProExpertProg** [comment]
Updated with note about fusing dynamic per-token quantization onto attention by using a similar approach to softmax computation

**Comment by ZJY0516** [comment]
May I ask why most fusion are limited to **Flashinfer (B200)**？It seems only nvfp4 is B200-specific.

**Comment by ProExpertProg** [comment]
@ZJY0516 that's where we get the fused kernels from. There's a section for non-B200 below that also asks for all-of-the-above fused kernels on other platforms. If you're interested in writing custom Triton kernels for example feel free to work on those.

**Comment by ZJY0516** [comment]
Thanks for explanation.

**Comment by nvpohanh** [comment]
- mla attention (prefill) + dynamic fp8 quant
- mla attention (prefill) + dynamic group fp8 quant

@ProExpertProg What is the difference between these two?

**Comment by ProExpertProg** [comment]
Dynamic quant is per-token, so we have to reduce across the whole hidden_size dimension to compute the scale. Dynamic group quant is per-group so we only reduce over a group (128 or 64 elements).

**Comment by ProExpertProg** [comment]
Yep MLA + dynamic group should be a priority! But some of this work is overlapping because MLA prefill is just MHA anyway, and MHA + dynamic per-token is more common than MHA + dynamic per-group

**Comment by ProExpertProg** [comment]
@andoorve I think with SP we actually fuse allgather to the following GEMM. I would be interested to see what the performance difference would be her, if rms_norm (+ quant) + all_gather would be better. So yeah definitely interested!

**Comment by LironKesem** [comment]
Hi, I want to take `mla_attention` with quantization using Helion. I want to make sure I understand which attention kernel is meant by 'mla_attention'

- Is it torch.ops.vllm.unified_mla_attention? If so, where can I find the implementation to inspect it?
- Or is it referring to standard [PyTorch nn.MultiheadAttention](https://docs.pytorch.org/docs/main/generated/torch.nn.MultiheadAttention.html)?
- regarding quant- is this fp8 or compressed datatype like MSFP (for the weights)

**Comment by ProExpertProg** [comment]
Another thing that would be good is a fused GEMM +collectives helion kernel we can use for platforms other than hopper. Currently we only have those kernels for hopper.

I think the helion repo already has a all gather + GEMM kernel; you can try integrating it into vLLM and measuring perf. Next step would be to add quantization support for it, as well as working in a GEMM + reduce scatter kernel.

**Comment by LironKesem** [comment]
@ProExpertProg 
issue: https://github.com/vllm-project/vllm/issues/30896

**Comment by banparth** [comment]
Hi @ProExpertProg, is there any "good first" subtask in this issue? I would like to contribute to that. Thanks!

**Comment by ProExpertProg** [comment]
@carlyou yeah feel free to take a look at MLA attn-quant fusions for MLA! If you could make a sub-issue I can assign you

**Comment by LeoYangXY** [comment]
Hi @ProExpertProg — I'd like to pick up the **Hopper Triton/Helion `all_reduce + rms_norm` (+ quant)** item from the "Other hardware" section.

Quick check from reading the thread: ROCm AITER is covered by #35712 / #35713 (cc @tjtanaa) — I don't see anything claiming the Hopper portable variant of `all_reduce + rms_norm`. Please correct me if I missed something.

**Hardware**: 2× H20 (sm_90, NV18 NVLink, FP8 native), so I can develop and benchmark TP=2 directly on target.

One questions before I open a sub-issue:
- **Triton-Distributed or Helion** for collective-bearing kernels? My understanding is Helion has no first-class distributed support yet, so I'd lean Triton-Distributed unless you say otherwise.

**Comment by rwtarpit** [comment]
@ProExpertProg can you add #41278 as sub-issue and assign it to me please.
thanks!!

**Comment by ProExpertProg** [comment]
@LeoYangXY I think Helion does support symmetric memory, although that will only work on CUDA. So feel free to use Triton-Distributed - that should support ROCm right? 

@rwtarpit done!

**Comment by LeoYangXY** [comment]
@ProExpertProg PR is up: #42230 — SymmMem-backed fused all_reduce + RMSNorm
  Triton kernel with static/dynamic-token/dynamic-group FP8 epilogues.

  This covers the "all_reduce + rms_norm (+ quant: all flavors)" item under
  "Other hardware (Hopper, AMD, etc.)". Tested on 4x H20 NVLink, 1.3–2.6×
  over NCCL + torch.nn.RMSNorm baseline depending on shape.

  Could you add a sub-issue for this and assign it to me?

  Two questions:
  1. Should I wire this into the `allreduce_rms_fusion` pass in the same PR,
     or keep that as a separate follow-up?
  2. Current speedups are 1.3–2.6× — good enough to land as-is, or prefer
     more tuning first?

**Comment by xsmccc** [comment]
Hi @ProExpertProg , I'm interested in working on the GEMM + static fp8 quant item from the
   "Other hardware" section.

  Regarding your question — "Can we do this with our existing cutlass kernel epilogues?" —
  I'd like to investigate this. My initial thinking:

  1. CUTLASS C3X epilogue path: Check if the SM120 blockwise FP8 GEMM dispatch supports a
  custom epilogue that applies static FP8 quant to the output. If feasible, this would be
  the least invasive approach and could benefit Hopper/SM120 alike.
  2. Triton GEMM path: If CUTLASS epilogue integration is too complex, write a Triton GEMM
  kernel with a fused FP8 quant epilogue. This is more portable (SM90/SM120/AMD) and aligns
   with the goal of avoiding platform-specific kernel explosion.
  3. Benchmark against Inductor: Either way, compare against Inductor-generated fused
  Triton to verify the fused kernel is actually faster.

  I have an RTX 5090 (SM120) for development, and would write the Triton path to be
  architecture-agnostic — I can validate on SM120 and CI can verify on other architectures.

  A bit about my background: I've been working on FP8 quantization and an inductor backend
  project involving operator fusion in C++. Happy to start with whichever direction you
  think is most useful.

  Does this sound reasonable? If so, I can create a sub-issue to track this.

**Comment by JediCoderX** [comment]
Hi @ProExpertProg , I'd like to pick up all_reduce + rms_norm + dynamic per-token fp8 quant from the "Needed (easy integration)" list. Plan is to extend allreduce_rms_fusion.py, following the same approach as the existing static fp8/nvfp4 patterns. Let me know if this is already in progress.
